dvc/repo/datasets.py

```diff
@@ -0,0 +1,350 @@
+from collections.abc import Iterator, Mapping
+from datetime import datetime
+from functools import cached_property
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Optional, Union, cast
+from urllib.parse import urlparse
+
+from attrs import Attribute, AttrsInstance, asdict, evolve, field, fields, frozen
+from attrs.converters import default_if_none
+
+from dvc.dvcfile import Lockfile, ProjectFile
+from dvc.exceptions import DvcException
+from dvc.log import logger
+from dvc_data.hashfile.meta import Meta
+
+if TYPE_CHECKING:
+    from dql.dataset import DatasetRecord  # type: ignore[import]
+    from typing_extensions import Self
+
+    from dvc.repo import Repo
+
+
+logger = logger.getChild(__name__)
+
+
+def default_str(v) -> str:
+    return default_if_none("")(v)
+
+
+def to_datetime(d: Union[str, datetime]) -> datetime:
+    return datetime.fromisoformat(d) if isinstance(d, str) else d
+
+
+def ensure(cls):
+    def inner(v):
+        return cls.from_dict(v) if isinstance(v, dict) else v
+
+    return inner
+
+
+class SerDe:
+    def to_dict(self: AttrsInstance) -> dict[str, Any]:
+        def filter_defaults(attr: Attribute, v: Any):
+            if attr.metadata.get("exclude_falsy", False) and not v:
+                return False
+            return attr.default != v
+
+        def value_serializer(_inst, _field, v):
+            return v.isoformat() if isinstance(v, datetime) else v
+
+        return asdict(self, filter=filter_defaults, value_serializer=value_serializer)
+
+    @classmethod
+    def from_dict(cls: type["Self"], d: dict[str, Any]) -> "Self":
+        _fields = fields(cast("type[AttrsInstance]", cls))
+        kwargs = {f.name: d[f.name] for f in _fields if f.name in d}
+        return cls(**kwargs)
+
+
+@frozen(kw_only=True)
+class DVCDataset(SerDe):
```

Comment:
All the `*Dataset` classes has a corresponding `*DatasetLock` class. `*Dataset` defines the specification for the dataset. `*DatasetLock` captures source information, but you'll see here that it also inherits from `*Dataset` classes. I am duplicating this information, because I want it to be possible to generate a `dvc.yaml` spec out of `lock`. This might be useful in the future. I definitely have wanted something similar with the stages (eg: `run cache` being useful to regenerate dvc.yaml etc.).

DatasetLock/Dataset are implemented from a serialization/deserialization perspective, and I haven't put much thought into that. We'll likely need to refactor them in the future. I do regret not using `pydantic` here, but `attrs` works for now.